### PR TITLE
Make NEST installation independent

### LINKED ISSUE #116: Test failure when NEST is installed system-wide
I have prepared a Debian packaging for NEST v2.8 targeting Debian proper (PR will come shortly). I have sorted out most things already. However, I am facing an issue when running the test suite for a NEST installation in /usr. This culprit is this:

```
Running test 'unittests/test_round_validate.sli'... 
   > Running  mpirun -np 1 /usr/bin/nest /usr/share/doc/nest/unittests/test_round_valid
ate.sli
   > NEST v2.8.0 (C) 2004 The NEST Initiative
   > 
   > Sep 30 17:34:23 file [Error]: FileOpenError
   >     Could not open the following file for writing: 
   >     "/usr/share/doc/nest/help/sli/array.hlp".
   > --------------------------------------------------------------------------
   > MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD 
   > with errorcode 126.
   > 
   > NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
   > You may or may not see output from other processes, depending on
   > exactly when Open MPI kills them.
   > --------------------------------------------------------------------------
   > --------------------------------------------------------------------------
   > mpirun has exited due to process rank 0 with PID 3952 on
   > node meiner exiting improperly. There are two reasons this could occur:
   > 
   > 1. this process did not call "init" before exiting, but others in
   > the job did. This can cause a job to hang indefinitely while it waits
   > for all processes to call "init". By rule, if one process calls "init",
   > then ALL processes must call "init" prior to termination.
   > 
   > 2. this process called "init", but exited without calling "finalize".
   > By rule, all processes that call "init" MUST call "finalize" prior to
   > exiting or it will be considered an "abnormal termination"
   > 
   > This may have caused other processes in the application to be
   > terminated by signals sent by mpirun (as reported here).
   > --------------------------------------------------------------------------
-> 126 (Failed: error in test script)
```

Is there a way to make this work -- maybe by writing into a temp dir? If not, I would exclude this test -- it is the only one that fails.

It would be nice to be able to run (a subset of) the tests without having to install NEST somewhere, but I could not figure out how -- it seems to heavily rely on its installation prefix.

Other open issues where I would appreciate you input are:
- is Python3 supported?
- if yes, is it possible to build for multiple Python versions with your autotools setup (doesn't look like it)
- can you confirm that all source code is (C) Nest initiative and GPL-2+? What about contributed pieces like ./testsuite/manualtests/stdp_prot.m ?

Thanks!

Sidenote: It would be nice, if the HEAD of master could be progressed to include the v2.8.0 tag.

### PR DESCRIPTION
**Pull request by tammoippen**
This PR makes a NEST installation independent from the source and build directory, as well as from the actual installation path. Further, it respects the standard GNU sub-installation directories and allows to change them on the command-line.

* `#define NEST_SOURCEDIR "@PKGSRCDIR@"` and `#define NEST_BUILDDIR "@PKGBUILDDIR@"` are removed, so NEST sources have no idea where to look for them.
* As #314 points out, there are at leaste two tests failing, if the sources are not at the expected place. 
    * `regressiontests/ticket-659-copyright.py` can now be deactivated by setting the environmental variable `NEST_SOURCE` to `"SKIP"`
    * the `validate` function does not regenerate the help files anymore for the test `unittests/test_round_validate.sli`
* #317 and #303 expect NEST to use standard GNU install directories, that are changeable during configuring. This is made possible by using the [GNUInstallDirs](https://cmake.org/cmake/help/v3.0/module/GNUInstallDirs.html) and replacing all nest specific install variables by the ones defined in there. During configuring, one can change the doc directory with `cmake .... -DCMAKE_INSTALL_DOCDIR=documentation/of/nest`
* If NEST is build dynamically, it uses relative [`RPATH`](https://cmake.org/Wiki/CMake_RPATH_handling#Recommendations) settings (`@loader_path` on OS X and `$ORIGIN` on Linux). This allows a NEST installation to be moved anywhere else. For NEST to find the sli init files, the help files and user modules (only PyNEST), some environmental variables have to be set, e.g. by sourcing `nest_vars.sh`.

This PR should fix #116, #303, #314 and #317. I propose @hanke, @sanjayankur31, @jougs and @heplesser  as reviewers.

### CODE REVIEW

File: examples/MyModule/CMakeLists.txt [hunk 1]
```diff
@@ -132,22 +155,39 @@ execute_process(
...
-
-# Complete paths of the install directories. Only used for the help installation.
-set( PKGSRCDIR ${PROJECT_SOURCE_DIR} )
-set( PKGBUILDDIR ${PROJECT_BINARY_DIR} )
-set( PKGDATADIR ${CMAKE_INSTALL_PREFIX}/${INSTALL_DATA_DIR} )
-set( PKGDOCDIR ${CMAKE_INSTALL_PREFIX}/${INSTALL_DOC_DIR} )
+# Install all stuff to NEST's install directories.
+set( CMAKE_INSTALL_LIBDIR ${NEST_LIBDIR}/nest CACHE STRING "object code libraries (lib/nest or lib64/nest or lib/<multiarch-tuple>/nest on Debian)" FORCE )
+set( CMAKE_INSTALL_DOCDIR ${NEST_DOCDIR} CACHE STRING "documentation root (DATAROOTDIR/doc/nest)" FORCE )
+set( CMAKE_INSTALL_DATADIR ${NEST_DATADIR} CACHE STRING "read-only architecture-independent data (DATAROOTDIR/nest)" FORCE )
```
Comment (heplesser) [inline examples/MyModule/CMakeLists.txt]: Is it intentional that these directories are relative paths? I suppose it is part of making nest relocatable?

File: extras/nest-config.in [hunk 2]
```diff
@@ -41,48 +45,43 @@ while test $# -gt 0; do
...
-        echo @CMAKE_CXX_COMPILER@
+        echo "@CMAKE_CXX_COMPILER@"
         ;;
 
     --static-libraries)
-        echo @static-libraries@
+        echo "@static-libraries@"
+        ;;
+    --docdir)
+        echo "@CMAKE_INSTALL_DOCDIR@"
```
Comment (heplesser) [inline extras/nest-config.in]: `--docdir` will print the relative path of the location of the documentation, not the absolute path. Is that sensible? If a user wants to know where the documentation is, the absolute path would be required, wouldn't it? The same applies to `--datadir` and `--libdir`.

File: examples/MyModule/CMakeLists.txt (hunk 1 above)
Comment (tammoippen) [inline examples/MyModule/CMakeLists.txt]: Yes. They should be relative paths. The full paths are created with `include( GNUInstallDirs )` and have the form `CMAKE_INSTALL_FULL_...`. The relative paths are used in the `install()` function as a destination within the `CMAKE_INSTALL_PREFIX`.

**Comment by heplesser** [comment]
@tammoippen NEST built and tests passed on my system without problems. I experimented with relocation and ran into some trouble, but that may be due to my stupidity:
- After moving the directory into which NEST was installed, I had to adjust `bin/nest_vars.sh` in that directory and then source that file into the shell to make the change effective. Is that right? Why did NEST work properly before relocating, even though I did not source `nest_vars.sh` then?
- I tried to build MyModule after relocating the install dir and after sourcing `nest_vars.sh`.  Cmake for MyModule still reported that it would install the module to `lib/nest` in the original install location, not in the new location. `$NEST_MODULE_PATH` contained the new location.
- `make installcheck` no longer works after relocation as it doesn't find `do_tests.sh`.

**Comment by tammoippen** [comment]
@heplesser 
> * After moving the directory into which NEST was installed, I had to adjust bin/nest_vars.sh in that directory and then source that file into the shell to make the change effective. Is that right? Why did NEST work properly before relocating, even though I did not source nest_vars.sh then?

The default directories (within `CMAKE_INSTALL_PREFIX`) are "hardcoded" into NEST. This is why NEST is working, if it is not relocated. When you relocate, you have to change the "hardcoded" values with environmental variables, as `nest_vars.sh` do.

> * I tried to build MyModule after relocating the install dir and after sourcing nest_vars.sh. Cmake for MyModule still reported that it would install the module to lib/nest in the original install location, not in the new location. $NEST_MODULE_PATH contained the new location.

You are right. `nest-config` is hardcoded to contain the `CMAKE_INSTALL_PREFIX`. I will have a look into this.

> * `make installcheck` no longer works after relocation as it doesn't find do_tests.sh.

`make` thinks, it installed everything to `CMAKE_INSTALL_PREFIX`. When relocating, make does not know, where to look for `do_tests.sh`. But you can check the installation with

    NEST_SOURCE=SKIP exec_prefix=<new-prefix> <new-prefix>/share/nest/extras/do_tests.sh --with-pynest

**Comment by heplesser** [comment]
@tammoippen Thanks for the information. The correct option for testing is `--test-pynest`, but things fail, since the path to the tests scripts seems to be fixed (I originally installed in `nest_tammo` and then renamed that dir `reloc`:
```
[plesser@Hanss-MacBook-Pro] ~/tmp (0)$ NEST_SOURCE=SKIP exec_prefix=/Users/plesser/tmp/reloc/ /Users/plesser/tmp/reloc/share/nest/extras/do_tests.sh --test-pynest

Phase 1: Testing if SLI can execute scripts and report errors
-------------------------------------------------------------
  Running test 'selftests/test_pass.sli'... Failed: unexpected exit code 126
Running  mpirun -np 1 /Users/plesser/tmp/reloc/bin/nest /Users/plesser/tmp/nest_tammo/share/doc/nest/selftests/test_pass.sli

Apr 25 15:37:57 file [Error]: FileOpenError
    Could not open the following file for reading: 
    "/Users/plesser/tmp/nest_tammo/share/doc/nest/selftests/test_pass.sli".
```

**Comment by tammoippen** [comment]
@heplesser Thank you. I will have a look.

**Comment by tammoippen** [comment]
@heplesser Please try again. 27f5c20 should fix the testing and ea22760 should fix the installation of MyModule. In both cases `nest_vars.sh` should be sourced.

**Comment by heplesser** [comment]
Testing after relocation works for SLI tests, but something strange happens for the PyNEST tests. When running `make installcheck`, 524 Python tests are run. When running after relocation as suggested above, only 241 are run. I have traced this back to `nosetests` being used when running via `make installcheck`, but the normal Python test harness while running `do_tests.sh` from the command line. I do not understand why this is the case.

Further, note that the following three lines in the installed `do_tests.sh` file still contain absolute locations:
```
TEST_BASEDIR=${NEST_DOC_DIR:-"/Users/plesser/tmp/nest_tammo/share/doc/nest"}

PYTHON_HARNESS="${NEST_DATA_DIR:-/Users/plesser/tmp/nest_tammo/share/nest}/extras/do_tests.py"

nosetests -v --with-xunit --xunit-file=${TEST_OUTDIR}/pynest_tests.xml ${NEST_PYTHON_PREFIX:-/Users/plesser/tmp/nest_tammo/lib/python2.7/site-packages}/nest/tests 2>&1 \
            | tee -a "${TEST_LOGFILE}" | grep -i --line-buffered "\.\.\. ok\|fail\|skip\|error" | sed 's/^/  /'
```

**Comment by heplesser** [comment]
Building MyModule after relocation works.

**Comment by heplesser** [comment]
The first line of `sli.el` also contains the original install path explicitly.

**Comment by heplesser** [comment]
It turned out that the problem with missing Python tests when running `do_tests.sh` form the command line instead of via `make installcheck` were due to the way we check for xunit support in nosetests. If `do_tests.sh` was run from a directory containing any failing Python tests, it would think that nosetests did not support xunit. This is fixed by #321.

File: extras/nest-config.in (hunk 2 above)
Comment (tammoippen) [inline extras/nest-config.in]: The arguments `nest-config --docdir` are used in MyModule to get the correct install paths (see your comment [here](https://github.com/nest/nest-simulator/pull/318#discussion_r60901809)). I think for most cases, these options are not relevant for users, and if they actually need them, they can prepend the result from `nest-config --prefix`. I will adjust the documentation.

**Comment by tammoippen** [comment]
@heplesser 
> Further, note that the following three lines in the installed `do_tests.sh` file still contain absolute locations:
```
TEST_BASEDIR=${NEST_DOC_DIR:-"/Users/plesser/tmp/nest_tammo/share/doc/nest"}
PYTHON_HARNESS="${NEST_DATA_DIR:-/Users/plesser/tmp/nest_tammo/share/nest}/extras/do_tests.py"
nosetests -v --with-xunit --xunit-file=${TEST_OUTDIR}/pynest_tests.xml ${NEST_PYTHON_PREFIX:-/Users/plesser/tmp/nest_tammo/lib/python2.7/site-packages}/nest/tests 2>&1 \
            | tee -a "${TEST_LOGFILE}" | grep -i --line-buffered "\.\.\. ok\|fail\|skip\|error" | sed 's/^/  /'
```

I used [parameter substitution](http://www.tldp.org/LDP/abs/html/parameter-substitution.html) here: `${NEST_DOC_DIR:-"/Users/plesser/tmp/nest_tammo/share/doc/nest"}` . So if `NEST_DOC_DIR` is set then use its value, otherwise use the previous install directory.

**Comment by heplesser** [comment]
@tammoippen Thanks for the shell-lesson :)!

Things work fine now, so :+1: from me.

**Comment by tammoippen** [comment]
Yes, `sli.el.in` has quite a lot references to `@CMAKE_INSTALL_FULL_DOCDIR@`, `@CMAKE_INSTALL_FULL_DATADIR@` and also to the sources (line 93) `@PROJECT_SOURCE_DIR@`. I think the Emacs support itself might need some polish: the access to the sources is necessary to get *SVN* information. The file `psvn.el` is an interface for the revision control tool subversion - we should "upgrade" to use the [Git supplied `git.el`](https://www.emacswiki.org/emacs/Git) extension. In 37b7077 I refactored these @- replacement into lisp variables, but I am not a regular emacs user / lisp programmer. Maybe someone with more experience can check this and do the polish. I will open an issue for for this.

**Comment by sanjayankur31** [comment]
It builds well here. I placed it in `/opt/` and everything seems to have gone into the right place. I built it with MPI and had to add the required snippet in my `~/.nestrc` before I ran `make installcheck` to run the tests. (Should nest have created a nestrc file for me when it ran the first time - I had one, but it was empty?)

When I was looking at `/opt/share/doc/nest/examples/nestrc.sli `, I noticed that the file still starts with:

```
% NEST has created this file for you from the template in
% <prefix>/share/doc/nest/examples/nestrc.sli.
```

Should `<prefix>` also be replaced with `CMAKE_INSTALL_PREFIX` here? Another `configure_file` directive in the `CMakeLists.txt` file? I'll let you know if anything else comes up. Thanks!

**Comment by tammoippen** [comment]
@sanjayankur31 
>  (Should nest have created a nestrc file for me when it ran the first time - I had one, but it was empty?)

Yes, NEST should have created a .nestrc file with some default content (on first execution) and it is required to edit the file for the `mpirun` command, if build with mpi. What do you mean with 'it was empty'? It should be a copy of `nestrc.sli`. Have you sourced `/opt/bin/nest_vars.sh` which sets some necessary environment variables for the relocated NEST.

> Should `<prefix>` also be replaced with CMAKE_INSTALL_PREFIX here?

I do not think, that this is necessary: the same .nestrc file can be used for multiple installations of nest and the `<prefix>` can change between them.

**Comment by sanjayankur31** [comment]
> Yes, NEST should have created a .nestrc file with some default content (on first execution) and it is required to edit the file for the mpirun command, if build with mpi. What do you mean with 'it was empty'? It should be a copy of nestrc.sli. Have you sourced /opt/bin/nest_vars.sh which sets some necessary environment variables for the relocated NEST.

Well, it was an empty file with nothing in it. Maybe I didn't source nest_vars. I'll look at it again later The rest looks good to me. :+1:

**Comment by heplesser** [comment]
Two thumbs up, so I will merge.

**Pull request merged**